### ISSUE TITLE
Off-by-one in number parsing

### ISSUE THREAD

**Issue by bjourne** [slavapestov/factor#1372]
```factor
most-positive-fixnum number>string string>number fixnum? .
t
```
That's good.
```factor
most-negative-fixnum number>string string>number fixnum? .
f
```
I think that expression should also be a fixnum then.

**Comment by jonenst** [comment]
That's because we accumulate the result as a positive number and neg at the end, and the most-negative-fixnum is 1 more than the most-positive-fixnum in absolute value. So this is the only fixnum for which it does that. Maybe change 'neg' or '*' to return a fixnum in this case ?

**Comment by jonenst** [comment]
```
M: bignum >integer
    dup most-negative-fixnum most-positive-fixnum between?
    [ >fixnum ] when ;

M: fixnum >integer ; inline

M: real >integer
    dup most-negative-fixnum most-positive-fixnum between?
    [ >fixnum ] [ >bignum ] if ; inline
```
So >integer can be called to ensure that bignums that fit become fixnums (ignoring the behavior with floats and ratios)

**Comment by bjourne** [comment]
I think adding that would be good even if it's not a huge bug. `number>string string>number` should as much as possible be a no-op.

**Comment by jonenst** [comment]
Also, this makes it "hard" to input the most-negative-fixnum because it's parsed as a bignum:
```factor
-576460752303423487 fixnum? ! t
-576460752303423488 >integer fixnum? ! t
-576460752303423488 fixnum? ! f
```

**Comment by jonenst** [comment]
This would work:
```diff
diff --git a/core/math/parser/parser.factor b/core/math/parser/parser.factor
index f5b73f8..e51b8b1 100644
--- a/core/math/parser/parser.factor
+++ b/core/math/parser/parser.factor
@@ -93,7 +93,12 @@ TUPLE: float-parse
     } cond ;
 
 : ?neg ( n/f -- -n/f )
-    [ neg ] [ f ] if* ; inline
+    [
+        dup bignum? [
+            dup first-bignum bignum=
+            [ drop most-negative-fixnum ] [ neg ] if
+        ] [ neg ] if
+    ] [ f ] if* ; inline
 
 : ?add-ratio ( m n/f -- m+n/f )
     dup ratio? [ + ] [ 2drop f ] if ; inline

```

**Comment by bjourne** [comment]
Why not just `>integer` like you suggested first?

**Comment by jonenst** [comment]
Because this code path is also used for floats and ratio.
Le 27 juin 2015 18:45, "Björn Lindqvist" <notifications@github.com> a
écrit :

**Comment by mrjbq7** [comment]
Fixed in 866b40271d774a0d4e5b9afc98acf39fdf834395.

@jonenst I'm using your patch, probably math.parser needs to be cleaned up at some point and you've already started that with your other patches.

Thanks!!
